graphcore_cloud_tools/paperspace_utils/symlink_datasets_and_caches.py

```diff
@@ -105,8 +129,374 @@ def run_symlinks(args):
+        s3_prefix = s3_response["Prefix"]
+        local_root = os.getenv(S3_DATASETS_DIR_ENV_VAR, DEFAULT_S3_DATASET_DIR)
+        for pre in s3_prefix.split("/"):
+            if pre not in local_root:
+                local_root = f"{local_root}/{pre}"
+        print(local_root)
+        if "/" != bucket_name[-1]:
+            bucket_name = f"{bucket_name}/"
+
+        def single_entry(s3_content_response: dict):
+            s3_object_name: str = s3_content_response["Key"]
+            full_s3file = f"{bucket_name}{s3_object_name}"
+            relative_file = s3_object_name.replace(s3_prefix, "").strip("/")
+            target = Path(local_root).resolve() / relative_file
+            return cls(
+                s3file=s3_object_name,
+                local_file=str(target),
+                size=s3_content_response.get("Size", 0),
+            )
+
+        return [single_entry(c) for c in s3_response["Contents"]]
+
+
+def list_files(client: "boto3.Client", dataset_name: str) -> List[GradientDatasetFile]:
+    dataset_prefix = f"{S3_DATASET_FOLDER}/{dataset_name}/"
+    out = client.list_objects_v2(Bucket="sdk", MaxKeys=10000, Prefix=dataset_prefix)
+    assert out["ResponseMetadata"].get("HTTPStatusCode", 200) == 200, "Response did not have HTTPS status 200"
+    assert not out["IsTruncated"], "Handling of truncated response is not handled yet"
+    if "Contents" not in out:
+        raise MissingDataset(f"Dataset '{dataset_name}' not found at 's3://sdk/{dataset_prefix}'")
+    logging.debug(f"S3 response {out}")
+    return GradientDatasetFile.from_response(out)
+
+
+def apply_symlink(
+    list_files: List[GradientDatasetFile], directory_map: Dict[str, List[str]]
+) -> List[GradientDatasetFile]:
+    def with_trailing_slash(path):
+        return path if path[-1] == "/" else f"{path}/"
+
+    source_target = {
+        with_trailing_slash(source): with_trailing_slash(target)
+        for target, sources in directory_map.items()
+        for source in sources
+    }
+    logging.debug(f"Mapping used for symling: {source_target}")
+    symlinked_list = []
+    for file in list_files:
+        local_file = file.local_file
+        for source, new_root in source_target.items():
+            if source in local_file:
+                local_file = local_file.replace(source, new_root)
+        symlinked_list.append(file._replace(local_file=local_file))
+    return symlinked_list
+
+
+class DownloadOutput(NamedTuple):
+    elapsed_seconds: float
+    gigabytes: float
+    error: Optional[Exception]
+
+
+def download_file_iterate_endpoints(aws_endpoints: List[str], *args, **kwargs) -> DownloadOutput:
+    # Randomly shuffles endpoints to load balance
+    aws_endpoints = aws_endpoints.copy()
+    random.shuffle(aws_endpoints)
+    error_in_loop = None
+    for aws_endpoint in aws_endpoints:
+        try:
+            return download_file(aws_endpoint, *args, **kwargs)
+        except Exception as error:
+            error_in_loop = error
```

Comment:
I've concatenated and added logging to make sure it is not lost, but I'm still raising from a single error at the end (because it needs to be a single exception)